src/cljam/io/sam/util/cigar.clj

```diff
@@ -83,20 +83,23 @@
 (defn decode-cigar-and-ref-length
   "Decode CIGAR string and length of alignment in reference.
   Returns a vector of [cigar, ref-length]."
-  [cigar-bytes]
+  [cigar-bytes l-seq]
   (let [buf (ByteBuffer/wrap cigar-bytes)
         sb (StringBuilder.)]
     (.order buf ByteOrder/LITTLE_ENDIAN)
-    (loop [ref-length 0]
+    (loop [ref-length 0 first-cigar-operation-clips-entire-read-flag false i 0]
       (if (.hasRemaining buf)
         (let [b (Integer/toUnsignedLong (.getInt buf))
               op (bit-and b 0xF)
               n  (unsigned-bit-shift-right b 4)]
           (doto sb
             (.append n)
             (.append (case op 0 \M 1 \I 2 \D 3 \N 4 \S 5 \H 6 \P 7 \= 8 \X)))
-          (recur (+ ref-length (case op 0 n 2 n 3 n 7 n 8 n 0))))
-        [(.toString sb) ref-length]))))
+          (recur (+ ref-length (case op 0 n 2 n 3 n 7 n 8 n 0))
+                 (if (and (= i 0) (= op 4) (= n l-seq))
+                   true first-cigar-operation-clips-entire-read-flag)
+                 (inc i)))
+        [(.toString sb) ref-length first-cigar-operation-clips-entire-read-flag]))))
```

Comment:
This function changed to do part of the determination of whether cigar has been replaced by a placeholder or not. This is because I thought it would perform better to make the decision at the same time as decoding.
Do you think this is good structure? Any ideas on how to write this function better?